packages/dashboard-core-plugins/src/WidgetLoaderPlugin.tsx

```diff
@@ -0,0 +1,94 @@
+import { useMemo, useCallback, type ComponentType, useEffect } from 'react';
+import type { ReactComponentConfig } from '@deephaven/golden-layout';
+import shortid from 'shortid';
+import {
+  assertIsDashboardPluginProps,
+  DashboardPluginComponentProps,
+  DehydratedDashboardPanelProps,
+  PanelEvent,
+  PanelOpenEventDetail,
+  LayoutUtils,
+  useListener,
+} from '@deephaven/dashboard';
+import { usePlugins } from '@deephaven/app-utils';
+import { isWidgetPlugin } from '@deephaven/plugin';
+
+export function WidgetLoaderPlugin(
+  props: DashboardPluginComponentProps
+): JSX.Element | null {
+  const plugins = usePlugins();
+  const supportedTypes = useMemo(() => {
+    const typeMap = new Map<string, ComponentType>();
+    plugins.forEach(plugin => {
+      if (!isWidgetPlugin(plugin)) {
+        return;
+      }
+
+      [plugin.supportedTypes].flat().forEach(supportedType => {
+        if (supportedType != null && supportedType !== '') {
+          typeMap.set(supportedType, plugin.component);
+        }
```

Comment:
Should we warn if there's already a type set for that type? Which plugin should take priority? Or should we allow multiple plugins to register and open panels in each of them (similar to JSON object viewer today)? We should be explicit about what we want the behaviour to be here, as right now it won't log anything and the last plugin registered will win (which may be fine).

I think I would probably stick with the behaviour you have, just add a comment/log about when it's overwriting an existing type. But you can have the final call.

Side note - should we allow patterns/regexes? I suppose doesn't have to be now, but the JSON object viewer plugin won't work with this. Which, again, is fine.

Reply:
Ya I've been thinking about that. I'm leaning towards something that only lets one plugin handle a widget. What would we expect to happen in Deephaven UI if there were 2 plugins that claim to handle a grid and we need to pick one?

The JSON object viewer can still work as a dashboard plugin with this.

I figure keep it simple w/ just strings for now and if there's a compelling use case for regex later we can add it.

I added a warning if 2 plugins register to handle the same widget type and the last plugin registered will be the plugin that handles the widget.

Reply:
Okay sounds good.